File: Data Structures & Algorithms/meeting-schedule/submission-1.cpp
/**
 * Definition of Interval:
 * class Interval {
 * public:
 * int start, end;
 * Interval(int start, int end) {
 * this->start = start;
 * this->end = end;
 * }
 * }
 */

class Solution {
public:
    bool canAttendMeetings(vector<Interval>& intervals) {
        // 1. Edge Case
        if (intervals.empty()) return true;

        // 2. Sort by Start Time
        // We need a custom comparator (lambda) because 'Interval' is a custom object.
        sort(intervals.begin(), intervals.end(), [](const Interval& a, const Interval& b) {
            return a.start < b.start;
        });

        // 3. Iterate and Check for Overlaps
        for (int i = 0; i < intervals.size() - 1; i++) {
            // Check if current meeting ends AFTER the next one starts
            if (intervals[i].end > intervals[i+1].start) {
                return false;
            }
        }

        return true;
    }
};

File: Data Structures & Algorithms/non-overlapping-intervals/submission-1.py
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0
        
        intervals.sort(key=lambda x: x[1])

        removed_count = 0
        current_end = intervals[0][1]

        for i in range(1, len(intervals)):
            interval = intervals[i]
            if interval[0] < current_end:
                removed_count += 1
            else:
                current_end = interval[1]
        
        return removed_count